llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -11510,12 +11522,27 @@ SDValue RISCVTargetLowering::lowerVECTOR_DEINTERLEAVE(SDValue Op,
     return DAG.getMergeValues(Res, DL);
   }
 
-  // TODO: Remove the e64 restriction once the fractional LMUL lowering
-  // is improved to always beat the vnsrl lowering below.
-  if (Subtarget.hasVendorXRivosVizip() && Factor == 2 &&
-      VecVT.getVectorElementType().getSizeInBits() == 64) {
+  if (Subtarget.hasVendorXRivosVizip() && Factor == 2) {
+    MVT VT = Op->getSimpleValueType(0);
     SDValue V1 = Op->getOperand(0);
     SDValue V2 = Op->getOperand(1);
+
+    // For fractional LMUL, check if we can use a higher LMUL
+    // instruction to avoid a vslidedown.
+    if (SDValue Src = foldConcatVector(V1, V2);
+        Src && getLMUL1VT(VT).bitsGT(VT)) {
+      EVT NewVT = VT.getDoubleNumVectorElementsVT();
```

Comment:
Is doubling the VT enough to prevent the vslidedown at mf4/mf8? Does NewVT need to be LMUL 1?

Reply:
It should be?  The original storage is still {LoSubVec, HiSubVec, tail elements}, so doubling gets everything in the non-tail.  I feel like I might be missing something in your question though, so if I am, feel free to push back.

Reply:
Ah I see that makes sense, we're doubling the VT to fit both the lo and hi in the one operand.